packages/typegpu/src/core/buffer/buffer.ts

```diff
@@ -83,6 +93,12 @@ export function isUsableAsVertex<T extends TgpuBuffer<AnyData>>(
 // Implementation
 // --------------
 
+type RestrictVertexUsages<TData extends AnyData> = TData extends
+  | { readonly type: 'loose-array' }
+  | { readonly type: 'loose-struct' }
+  ? 'vertex'[]
+  : ('uniform' | 'storage' | 'vertex')[];
```

Comment:
Tiny nit, can be made a bit more general, in case we want to add more loose schemas.

```suggestion
type RestrictVertexUsages<TData extends AnyData> = TData extends
  | { readonly type: WgslTypeLiteral }
  ? ('uniform' | 'storage' | 'vertex')[]
  : 'vertex'[];
```